pkg/server/auth/auth.go

```diff
@@ -198,29 +198,58 @@ func WithAPIAuth(next http.Handler, srv *AuthServer, publicRoutes []string) http
 		}
 	}
 
-	return http.HandlerFunc(func(rw http.ResponseWriter, r *http.Request) {
-		if IsPublicRoute(r.URL, publicRoutes) {
-			next.ServeHTTP(rw, r)
-			return
-		}
+	return &authenticatedMiddleware{
+		next:            next,
+		srv:             srv,
+		publicRoutes:    publicRoutes,
+		principalGetter: multi,
+		locks:           newRefreshLocker(),
+	}
+}
+
+type authenticatedMiddleware struct {
+	srv             *AuthServer
+	publicRoutes    []string
+	next            http.Handler
+	principalGetter PrincipalGetter
+	locks           *refreshLocks
+}
 
-		principal, err := multi.Principal(r)
-		if err != nil || principal == nil {
-			var refreshErr error
-			principal, refreshErr = srv.Refresh(rw, r)
-			if refreshErr != nil || principal == nil {
-				srv.Log.V(logger.LogLevelWarn).Info("refreshing token failed", "err", refreshErr, "principal", principal)
-				srv.Log.V(logger.LogLevelWarn).Info("Authentication failed", "err", err, "principal", principal)
+func getRefreshToken(r *http.Request) string {
+	refreshTokenCookie, err := r.Cookie(RefreshTokenCookieName)
+	if err != nil {
+		return ""
+	}
 
-				JSONError(srv.Log, rw, "Authentication required", http.StatusUnauthorized)
-				return
-			}
+	return refreshTokenCookie.Value
+}
+
+func (a *authenticatedMiddleware) ServeHTTP(rw http.ResponseWriter, r *http.Request) {
+	if IsPublicRoute(r.URL, a.publicRoutes) {
+		a.next.ServeHTTP(rw, r)
+		return
+	}
+
+	principal, err := a.principalGetter.Principal(r)
+
+	if refreshToken := getRefreshToken(r); err != nil || principal == nil && refreshToken != "" {
```

Comment:
I think we still wanna get into this block if refreshToken is empty, the `"Authentication required"`  inside this block is sort of the global 401

Reply:
Slightly worrying that no tests fail in this case :-|